### PR TITLE
feat: Hash assets before CLI upload

### PR DESCRIPTION
**Pull request by etnoy**
This PR adds support for client-side hashing for the Immich CLI.

* Performs sha1 hash of all assets and checks them with the new API endpoint before uploading
* Removes the device uuid option as it's no longer used
* Adds a --skip-hash option to skip hashing and just upload it all, leaving deduplication to the server
* Adds some more error messages to the output

TODO:
* Neater progress bar for hashing

### CODE REVIEW

File: bin/index.ts [hunk 1]
```diff
@@ -175,54 +168,96 @@ async function upload(
...
-        const fileStat = fs.statSync(filePath);
+  const crypto = require('crypto');
+
+  const sha1 = (filePath: string) =>
+    new Promise<string>((resolve, reject) => {
+      const hash = crypto.createHash('sha1');
+      const rs = fs.createReadStream(filePath);
+      rs.on('error', reject);
+      rs.on('data', (chunk) => hash.update(chunk));
+      rs.on('end', () => resolve(hash.digest('hex')));
```
Comment (jrasm91) [inline bin/index.ts:200]: Thoughts on changing this to base64 instead of hex?

Comment (etnoy) [reply bin/index.ts:200]: Great idea. We should switch to base64 here and in the bulk upload check controller on the server

Comment (jrasm91) [reply bin/index.ts:200]: I updated the server to accept either hex or base64 :+1:, so I don't think it matters too much either way here. `base64` is more compact though.

**Comment by alextran1502** [comment]
Can you help with a procedure for how to test this PR?

**Comment by etnoy** [comment]
> Can you help with a procedure for how to test this PR?

I've simply been testing with a folder with a bunch of files in them, uploading and re-uploading them to see what happens in different scenarios

**Comment by Orbitally** [comment]
Just thought I'd pop in to say that I just gave this a shot and it worked great! I had uploaded most of my assets with the current release of the CLI and running this version it flagged all the previously uploaded assets as being duplicates and didn't do anything with them while uploading the two new pictures I had added to the source folder to the server.

**Comment by etnoy** [comment]
Closing this as this CLI is about to be deprecated. Use the CLI in the main repo (beta) for hashing support.

**Pull request closed without merging**